quickwit/quickwit-search/src/leaf.rs

```diff
@@ -879,6 +877,130 @@ impl CanSplitDoBetter {
         }
     }
 
+    /// This function tries to detect upfront which splits contain the top n hits and convert other
+    /// split searches to count only searches. It also optimizes split order.
+    ///
+    /// Returns the search_requests with their split.
+    fn optimize(
+        &self,
+        request: Arc<SearchRequest>,
+        mut splits: Vec<SplitIdAndFooterOffsets>,
+    ) -> Result<Vec<(SplitIdAndFooterOffsets, SearchRequest)>, SearchError> {
+        self.optimize_split_order(&mut splits);
+        // TODO: we maybe want here some deduplication + Cow logic
+        let mut split_with_req = splits
+            .into_iter()
+            .map(|split| (split, (*request).clone()))
+            .collect::<Vec<_>>();
+
+        if request.aggregation_request.is_some() || request.search_after.is_some() {
+            return Ok(split_with_req);
+        }
+        let query_ast: QueryAst = serde_json::from_str(request.query_ast.as_str())
+            .map_err(|err| SearchError::InvalidQuery(err.to_string()))?;
+
+        // TODO: Update the logic to handle start_timestamp end_timestamp ranges
+        if request.start_timestamp.is_some() || request.end_timestamp.is_some() {
+            return Ok(split_with_req);
+        }
+
+        if !matches!(query_ast, QueryAst::MatchAll) {
+            return Ok(split_with_req);
+        }
+
+        // Count the number of splits which contain enough documents
+        let count_required_splits =
+            move |split_with_req: &[(SplitIdAndFooterOffsets, SearchRequest)],
+                  num_requested_docs: u64| {
+                let mut num_docs = 0;
+                split_with_req
+                    .iter()
+                    .take_while(|(split, _req)| {
+                        let need_more_docs = num_docs < num_requested_docs;
+                        num_docs += split.num_docs;
+                        need_more_docs
+                    })
+                    .count()
+            };
+
+        // reuse the detected sort order in split_filter
+        // we want to detect cases where we can convert some split queries to count only queries
+        let num_requested_docs = request.start_offset + request.max_hits;
+        match self {
+            CanSplitDoBetter::SplitIdHigher(_) => {
```

Comment:
why do we apply the logic on `CanSplitDoBetter` instead of the actual sort order here?
Can't we have CanSplitDoBetter set to informative even though docs are requested to be ordered by timestamp?

Reply:
ah answering my own question. if sorted by timestmap, but we don't have a bound, then `CanSplitDoBetter` is set to Higher or Lower but with `None`